### PR TITLE
lava and weather immunities refactor (also jump boots fix)

### LINKED ISSUE #34257: Jump boots set you on fire when jumping over lava
When you jump over lava it still burns you, even when you wouldn't land on lava

I don't know the round ID or remember anything about the round because this was a while ago

To reproduce: get jump boots, stand next to a lava river small enough that you'll land on ground, face lava river, use jump boots (recommended but not part of reproduction steps: use extinguisher and burn patch after you get set on fire)

Might be a bug, might be intentional, but you should really be able to jump over lava

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
In remembrance of all those people who used jump boots to cross lava unaware of an issue c*ders wouldn't fix....

EDIT: This is now a lava and weather immunities refactor:
Weather immunities are now status traits since they have a multitude of sources (especially for lava) which might conflict with one another otherwise.
The lava burn_stuff proc has also been been refactored in different procs, mostly because of that snowdin subtype with inconsistent, old checks.
Weather datums should now use `can_weather_act` instead of `weather_act` to check if something can be affected by weather or not, as they should.
All movables can protect contained mobs if they have the relative weather immunity traits. This works at any contents depth.
No more snowflake `weather_protection` variable for closets.
Removed the weather_immunities list from living mobs (simple animals still have it but it's only for traits assignment on init because way too many child types lack the immunities of their parents).
Removed some unused defines.
Renamed some variables as per guidelines.
It has been tested.
And yea, jump boots fixed because that's the original scope of this PR.

## Why It's Good For The Game
This will fix #34257 and improve the code a fair bit.

## Changelog
:cl:
fix: fixes people catching fire when using jump or rocket boots to leap over a gap of lava.
refactor: refactored lava burn and weather immunities code. 
/:cl:

### CODE REVIEW

**Comment by AMonkeyThatCodes** [comment]
this is a balance change and has been purposefully been kept in to prevent gulagged people from just getting over with ease.

**Comment by LemonInTheDark** [comment]
https://github.com/tgstation/tgstation/issues/25858#issuecomment-334271477

This should be a change on jump boots, not a change on lava. Assistant A throwing assistant B out of perma is dumb

**Comment by Ghommie** [comment]
> 
> 
> this is a balance change and has been purposefully been kept in to prevent gulagged people from just getting over with ease.

Oh dang you are right. wait a moment.

**Comment by Couls** [comment]
if you plan on changing it on the jumpboots, take a look at this PR https://github.com/tgstation/tgstation/pull/34279

**Comment by Ghommie** [comment]
> 
> 
> if you plan on changing it on the jumpboots, take a look at this PR #34279

I'm glad I refactored movetype code last year so we wouldn't end up doing these sort of mistakes again.

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 1]
```diff
@@ -8,7 +8,6 @@
 	pass_flags = PASSTABLE | PASSMOB
 	mob_size = MOB_SIZE_TINY
 	desc = "A generic pAI mobile hard-light holographics emitter. It seems to be deactivated."
-	weather_immunities = list(WEATHER_ASH)
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/mob/living/silicon/pai/pai.dm:11]: Same value as parent really.

File: code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm [hunk 2]
```diff
@@ -48,15 +48,15 @@
 		//Because the leaping sprite is bigger than the normal one
 		body_position_pixel_x_offset = -32
 		body_position_pixel_y_offset = -32
-		LAZYADD(weather_immunities, WEATHER_LAVA)
 		update_icons()
+		ADD_TRAIT(src, TRAIT_MOVE_FLOATING, LEAPING_TRAIT) //Throwing itself doesn't protect mobs against lava (because gulag).
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm:52]: Not really. It's just stupid that benos use lava immunity for leaping while jump boots use a movetype imo (also the cold plasma lava tiles of snowdin use different immunities)

**Pull request merged**